src/main/java/uk/gov/justice/hmpps/casenotes/alertbackfill/AlertCaseNoteVerification.kt

```diff
@@ -0,0 +1,146 @@
+import com.sun.org.apache.xalan.internal.lib.ExsltDatetime.date
+import jakarta.annotation.PostConstruct
+import org.springframework.stereotype.Service
+import uk.gov.justice.hmpps.casenotes.alertbackfill.ActiveInactive.ACTIVE
+import uk.gov.justice.hmpps.casenotes.alertbackfill.ActiveInactive.INACTIVE
+import uk.gov.justice.hmpps.casenotes.config.ServiceConfig
+import uk.gov.justice.hmpps.casenotes.domain.Note
+import uk.gov.justice.hmpps.casenotes.domain.NoteRepository
+import uk.gov.justice.hmpps.casenotes.domain.SubType
+import uk.gov.justice.hmpps.casenotes.domain.SubTypeRepository
+import uk.gov.justice.hmpps.casenotes.domain.System
+import uk.gov.justice.hmpps.casenotes.domain.TypeKey
+import uk.gov.justice.hmpps.casenotes.domain.createdBetween
+import uk.gov.justice.hmpps.casenotes.domain.matchesOnType
+import uk.gov.justice.hmpps.casenotes.domain.matchesPersonIdentifier
+import uk.gov.justice.hmpps.casenotes.events.AdditionalInformation
+import uk.gov.justice.hmpps.casenotes.events.DomainEvent
+import uk.gov.justice.hmpps.casenotes.integrations.ManageUsersService
+import uk.gov.justice.hmpps.casenotes.integrations.UserDetails
+import java.time.LocalDate
+import java.time.temporal.ChronoUnit
+
+@Service
+class AlertCaseNoteVerification(
+  private val manageUsersService: ManageUsersService,
+  private val alertService: AlertService,
+  private val noteRepository: NoteRepository,
+  private val subTypeRepository: SubTypeRepository,
+  private val telemetryClient: TelemetryClient,
+  private val serviceConfig: ServiceConfig,
+) {
+  fun verify(domainEvent: DomainEvent<AlertVerificationInformation>) {
+    setDpsUserIfNotSet()
+    val info = domainEvent.additionalInformation
+    val alerts = alertService.getAlertsOfInterest(info.personIdentifier, info.from, info.to).content
+    val existingNotes = noteRepository.findAll(info.asSpecification())
+    val toCreate = alerts.flatMap {
+      val matches = existingNotes matching it
+      buildList {
+        if (matches[ACTIVE] == null) add(ACTIVE to it)
+        if (it.madeInactive() && matches[INACTIVE] == null) add(INACTIVE to it)
+      }
+    }.groupBy({ it.first }, { it.second })
+
+    if (toCreate.isEmpty()) return
+    telemetryClient.trackEvent("MissingAlertCaseNotes", toCreate.properties(), mapOf())
+
+    if (serviceConfig.actionMissingCaseNotes) {
+      val subTypes = subTypeRepository.findByKeyIn(
+        setOf(TypeKey("ALERT", "ACTIVE"), TypeKey("ALERT", "INACTIVE")),
+      ).associateBy { ActiveInactive.valueOf(it.code) }
+      val notes = toCreate.flatMap { e ->
+        e.value.map { it.toNote(info.personIdentifier, e.key, subTypes) }
+      }
+      noteRepository.saveAll(notes)
+    }
+  }
+
+  private fun AlertVerificationInformation.asSpecification() =
+    matchesPersonIdentifier(personIdentifier)
+      .and(matchesOnType(true, mapOf("ALERT" to setOf())))
+      .and(createdBetween(from.atStartOfDay(), to.plusDays(1).atStartOfDay(), true))
```

Comment:
Guess the thinking here is anything created at a minute to midnight is functionally a different note? Would it make any sense to do +- 1-12 hours or something like that?